### PR TITLE
Add context to the Copilot Chat

### PR DESCRIPTION
**Pull request by lukka**
@benmcmorran @spebl @hnrqbaggio FYI

### CODE REVIEW

File: Extension/src/LanguageServer/client.ts [hunk 1]
```diff
@@ -2187,6 +2197,12 @@ export class DefaultClient implements Client {
         await this.languageClient.sendNotification(DidOpenNotification, params);
     }
 
+    public async getChatContext(): Promise<ChatContextResult | undefined> {
+        await this.ready;
+        const result: ChatContextResult | undefined = await this.languageClient.sendRequest(CppContextRequest, null);
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: Is there a reason for the result variable versus just returning it?

**Review by sean-mcmanus — DISMISSED** [review]

File: Extension/src/LanguageServer/extension.ts [hunk 2]
```diff
@@ -246,6 +246,77 @@ export async function activate(): Promise<void> {
...
+    await setupCppChatVariable(util.extensionContext);
+}
+
+export async function setupCppChatVariable(context: vscode.ExtensionContext | undefined): Promise<void>
+{
+    if (!context) {
+        return;
+    }
+
+    vscode.chat.registerChatVariableResolver('cpp',
```
Comment (benmcmorran) [inline Extension/src/LanguageServer/extension.ts:259]: Will we have a way to measure usage of `#cpp`? I'm wondering if we should add our own telemetry, or if VS Code will emit some automatically.

File: Extension/package.json [hunk 3]
```diff
@@ -38,7 +38,9 @@
         "Snippets"
     ],
     "enabledApiProposals": [
-        "terminalDataWriteEvent"
+        "terminalDataWriteEvent",
+        "chatParticipant",
+        "chatVariableResolver"
```
Comment (bobbrow) [inline Extension/package.json:43]: We will need approval to use these APIs before this PR can be approved.

**Review by bobbrow — CHANGES_REQUESTED** [review]
(not actually requesting changes, just blocking the PR until we have approval for the API proposals)

File: Extension/src/LanguageServer/client.ts (hunk 1 above)
Comment (lukka) [reply Extension/src/LanguageServer/client.ts]: removed the local var.

File: Extension/src/LanguageServer/extension.ts (hunk 2 above)
Comment (lukka) [reply Extension/src/LanguageServer/extension.ts:259]: usage telemetry added

File: Extension/src/LanguageServer/extension.ts [hunk 4]
```diff
@@ -246,6 +246,78 @@ export async function activate(): Promise<void> {
...
+
+                const language = fixUpLanguage(chatContext.language);
+                const compiler = fixUpCompiler(chatContext.compiler);
+                const standardVersion = fixUpStandardVersion(chatContext.standardVersion);
+                const targetPlatform = fixTargetPlatform(chatContext.targetPlatform);
+                const value = `I am working on a project of the following nature:
+- Using ${language}. Prefer a solution written in ${language} to any other language. Call out which language you are using in the answer.
+- Using the ${language} standard language version ${standardVersion}. Prefer solutions using the new and more recent features introduced in ${standardVersion}. Call out which standard version you are using in the answer.
+- Using the ${compiler} compiler. Prefer solutions supported by the ${compiler} compiler.
+- Targeting the ${targetPlatform} platform. Prefer solutions and API that are supported on ${targetPlatform} .
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/extension.ts]: There's an extra space here -- not sure if that matters though: `${targetPlatform} .`

Comment (lukka) [reply Extension/src/LanguageServer/extension.ts]: fixed, thank you!

File: Extension/src/LanguageServer/extension.ts [hunk 5]
```diff
@@ -246,6 +246,78 @@ export async function activate(): Promise<void> {
...
+}
+
+export async function setupCppChatVariable(context: vscode.ExtensionContext | undefined): Promise<void>
+{
+    if (!context) {
+        return;
+    }
+
+    vscode.chat.registerChatVariableResolver('cpp',
+        `Describes the the C++ language features that can be used according
```
Comment (jogo-) [inline Extension/src/LanguageServer/extension.ts:260]: duplicated word "the the"

**Comment by sean-mcmanus** [comment]
@lukka Did you want to get these changes into our next release 1.20.2?

**Comment by lukka** [comment]
> @lukka Did you want to get these changes into our next release 1.20.2?

Not yet, thank you for the heads up!

**Comment by lukka** [comment]
closing in favor of #12577

**Pull request closed without merging**